packages/ai-jsx/src/batteries/docs.tsx

```diff
@@ -1,154 +1,503 @@
-}
+  export class Partition {
+    constructor(readonly name: string, readonly firstPageToken?: string) {
+      /** The unique name of this partition. */
+      this.name = name;
+      /** An optional token for requesting the first page. */
+      this.firstPageToken = firstPageToken;
+    }
+  }
 
-/**
- * A vector database that stores vectors in memory. In-memory is nice for rapid prototyping. However, it'll be slower, because every run needs to populate the database. And it's limited by the current process' memory.
- */
-export class DefaultInMemoryVectorStore<Metadata extends Jsonifiable = Jsonifiable> implements VectorStore {
-  private constructor(private readonly langChainVectorStore: MemoryVectorStore) {}
+  /** A function responsible for loading and parsing a corpus. */
+  export type Loader<DocumentMetadata extends Jsonifiable = Jsonifiable> = (
+    request: Request
+  ) => Promise<Response<Document<DocumentMetadata>>>;
 
-  static async fromDocuments<Metadata extends Jsonifiable = Jsonifiable>(
-    docs: Document<Metadata>[],
-    embeddings: Embeddings
-  ) {
-    const lcMemoryStore = await MemoryVectorStore.fromDocuments(
-      docs.map((doc) => toLangChainDoc(doc)),
-      embeddings
-    );
-    return new this<Metadata>(lcMemoryStore);
+  /** A function responsible for loading a corpus when parsing is handled separately. */
+  export type RawLoader = (request: Request) => Promise<Response<RawDocument>>;
+
+  export function toLoader<DocumentMetadata extends Jsonifiable = Jsonifiable>(
+    rawLoader: RawLoader,
+    parser: Parser<DocumentMetadata> = defaultParser
+  ): Loader<DocumentMetadata> {
+    return async (request: Request) => {
+      const response = await rawLoader(request);
+      if (response.page) {
+        const documents = await Promise.all(response.page.documents.map(parser));
+        return new Response(new Page(documents, response.page.nextPageToken), response.partitions);
+      }
+      return new Response<Document<DocumentMetadata>>(undefined, response.partitions);
+    };
   }
 
-  async search<Filter = never>(
-    query: string,
-    opts: { filter?: Filter | undefined; limit: number }
-  ): Promise<VectorSearchResult<Metadata>[]> {
-    const docs = await this.langChainVectorStore.similaritySearchWithScore(query, opts.limit);
-    return docs.map(([doc, score]) => ({
-      document: doc as Document<Metadata>,
-      score,
-    }));
+  class StaticLoader<DocumentMetadata extends Jsonifiable = Jsonifiable> {
+    constructor(readonly documents: Document<DocumentMetadata>[], readonly pageSize?: number) {}
+
+    load(request: Request): Promise<Response<Document<DocumentMetadata>>> {
+      if (this.pageSize) {
+        const page: number = Number(request.pageToken ?? 0);
+        const start = page * this.pageSize;
+        const end = start + this.pageSize;
+        if (end > this.documents.length) {
+          return Promise.resolve(new Response(new Page(this.documents.slice(start))));
```

Comment:
```suggestion
          return Promise.resolve(new Response(new Page([...documents])));
```

If you're just trying to clone the array, this is a cleaner way to do it IMO.

Reply:
It's not a clone, it's a sublist.  Am I missing something?

Reply:
oh whoops no, I was misreading `start` as `0`.